Dockerfile.rocm

```diff
@@ -70,16 +73,33 @@ RUN if [ "$BUILD_FA" = "1" ]; then \
     && cd ..; \
     fi
 
-COPY ./ /app/vllm
-
-RUN python3 -m pip install --upgrade pip
-RUN python3 -m pip install xformers==0.0.23 --no-deps
-
 # Error related to odd state for numpy 1.20.3 where there is no METADATA etc, but an extra LICENSES_bundled.txt.
 # Manually removed it so that later steps of numpy upgrade can continue
 RUN if [ "$BASE_IMAGE" = "rocm/pytorch:rocm6.0_ubuntu20.04_py3.9_pytorch_2.1.1" ]; then \
     rm -rf /opt/conda/envs/py_3.9/lib/python3.9/site-packages/numpy-1.20.3.dist-info/; fi
 
+# build cupy
+RUN if [ "$BUILD_CUPY" = "1" ]; then \
+    mkdir -p libs \
+    && cd libs \
+    && git clone -b hipgraph_enablement --recursive https://github.com/ROCm/cupy.git \
+    && cd cupy \
+    && pip install mpi4py-mpich \
```

Comment:
Do we need MPI for CuPy? For NVIDIA GPUs, we use TCP store instead of MPI.

Reply:
vLLM uses a hack that terminates the TCP store used by cupy right after the cupy nccl backend is initialized:

https://github.com/vllm-project/vllm/blob/29e70e3e88698feca9509cf07fcf06b12163f1c3/vllm/model_executor/parallel_utils/cupy_utils.py#L93-L96

I did this because I found that otherwise the worker processes hang when they are terminated. If ROCm cupy uses MPI, then vLLM might need a similar hack to prevent deadlocks at termination time.

Reply:
Hi @WoosukKwon we prefer to use MPI for ROCm Cupy. Is there any specific reason to choose TCP Store instead of MPI from the vLLM side?

Reply:
> vLLM uses a hack that terminates the TCP store used by cupy right after the cupy nccl backend is initialized:
> 
> https://github.com/vllm-project/vllm/blob/29e70e3e88698feca9509cf07fcf06b12163f1c3/vllm/model_executor/parallel_utils/cupy_utils.py#L93-L96
> 
> I did this because I found that otherwise the worker processes hang when they are terminated. If ROCm cupy uses MPI, then vLLM might need a similar hack to prevent deadlocks at termination time.

@WoosukKwon (1) Can you give more context about this deadlock issue when the processes are terminated? We will need to test it to see whether we see the deadlock also to make sure whether we need this "stop" hack. (2) What the reason the TCPstore is used instead of MPI? is there any performance issue with MPI? As @lcskrishna mentioned, we have tested more on the MPI path.

Reply:
Hi @lcskrishna @hongxiayang , we used TCP store just to reduce dependencies. We can use MPI if you prefer it over TCP store. Although I haven't tested MPI + CuPy on NVIDIA GPUs, I believe it works.

The deadlock issue is that, when the main process is terminated, the process hangs waiting for other processes spawned by cupy TCP store. The `_NCCL_BACKEND._store.stop()` hack is to avoid this.

```
Traceback (most recent call last):
  File "/usr/lib/python3.10/multiprocessing/util.py", line 357, in _exit_function
Process ExceptionAwareProcess-1:
    p.join()
  File "/usr/local/lib/python3.10/dist-packages/cupyx/distributed/_store.py", line 38, in join
    super().join()
  File "/usr/lib/python3.10/multiprocessing/process.py", line 149, in join
    res = self._popen.wait(timeout)
  File "/usr/lib/python3.10/multiprocessing/popen_fork.py", line 43, in wait
    return self.poll(os.WNOHANG if timeout == 0.0 else 0)
  File "/usr/lib/python3.10/multiprocessing/popen_fork.py", line 27, in poll
    pid, sts = os.waitpid(self.pid, flag)
KeyboardInterrupt: 
Traceback (most recent call last):
  File "/usr/lib/python3.10/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/usr/local/lib/python3.10/dist-packages/cupyx/distributed/_store.py", line 32, in run
    super().run()
  File "/usr/lib/python3.10/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.10/dist-packages/cupyx/distributed/_store.py", line 87, in _server_loop
    c_socket, addr = s.accept()
  File "/usr/lib/python3.10/socket.py", line 293, in accept
    fd, addr = self._accept()
```

I'm not sure whether this also happens for the MPI backend. Could you please test it out and see whether it happens?

Reply:
> Hi @lcskrishna @hongxiayang , we used TCP store just to reduce dependencies. We can use MPI if you prefer it over TCP store. Although I haven't tested MPI + CuPy on NVIDIA GPUs, I believe it works.
> 
> The deadlock issue is that, when the main process is terminated, the process hangs waiting for other processes spawned by cupy TCP store. The `_NCCL_BACKEND._store.stop()` hack is to avoid this.
> 
> ```
> Traceback (most recent call last):
>   File "/usr/lib/python3.10/multiprocessing/util.py", line 357, in _exit_function
> Process ExceptionAwareProcess-1:
>     p.join()
>   File "/usr/local/lib/python3.10/dist-packages/cupyx/distributed/_store.py", line 38, in join
>     super().join()
>   File "/usr/lib/python3.10/multiprocessing/process.py", line 149, in join
>     res = self._popen.wait(timeout)
>   File "/usr/lib/python3.10/multiprocessing/popen_fork.py", line 43, in wait
>     return self.poll(os.WNOHANG if timeout == 0.0 else 0)
>   File "/usr/lib/python3.10/multiprocessing/popen_fork.py", line 27, in poll
>     pid, sts = os.waitpid(self.pid, flag)
> KeyboardInterrupt: 
> Traceback (most recent call last):
>   File "/usr/lib/python3.10/multiprocessing/process.py", line 314, in _bootstrap
>     self.run()
>   File "/usr/local/lib/python3.10/dist-packages/cupyx/distributed/_store.py", line 32, in run
>     super().run()
>   File "/usr/lib/python3.10/multiprocessing/process.py", line 108, in run
>     self._target(*self._args, **self._kwargs)
>   File "/usr/local/lib/python3.10/dist-packages/cupyx/distributed/_store.py", line 87, in _server_loop
>     c_socket, addr = s.accept()
>   File "/usr/lib/python3.10/socket.py", line 293, in accept
>     fd, addr = self._accept()
> ```
> 
> I'm not sure whether this also happens for the MPI backend. Could you please test it out and see whether it happens?

@WoosukKwon Quick question for verification: Regarding "when the main process is terminated"? do you mean it was killed manually before it completes in running throughput benchmarking script?

Reply:
@hongxiayang Not really. Without the two-line hack on cupy TCP store, the process hangs when it normally terminates (e.g., after running `python examples/llm_engine_example.py -tp 2`).

Reply:
> @hongxiayang Not really. Without the two-line hack on cupy TCP store, the process hangs when it normally terminates (e.g., after running `python examples/llm_engine_example.py -tp 2`).

@WoosukKwon It seem that we do not need to do anything for this situation.
(1) I tested the `examples/llm_engine_example.py -tp 2`, it completed fine without deadlock or hang without any patch to call mpi_comm.Abort(). (2) I also tested with throughput benchmarking script and press Ctrl-C in the middle of the script, the script stopped cleanly. (3) I discussed with @lcskrishna earlier, and he also thought that mpi might not need any additional hack.

Reply:
Got it. Thanks for testing!